### DISCUSSION TITLE
gcc 11.3 loader error (also gcc 12.1)

### DISCUSSION THREAD

**Discussion by sanjayg0** [Homebrew/discussions#3296] [Everyday usage]
Upon upgrading to GCC to 11.3 (from 11.2), I am experiencing a loader error:

```
ld: in section __TEXT,__text reloc 490: symbol index out of range file 'sld2d1u.o' for architecture arm64
collect2: error: ld returned 1 exit status
```

I looked in the .s file for the subroutine for __TEXT and see the following but I don't know enough to be able to figure out what is going on (making a standalone example displaying the same error will be quite hard).

```
LEFDE0:
        .section __TEXT,__eh_frame,coalesced,no_toc+strip_static_syms+live_support
EH_frame1:
        .set L$set$61,LECIE1-LSCIE1
        .long L$set$61
```

Note, the same error occurs with gcc@12 (gcc 12.1).

(1) Any hints on how to chase this down?
(2) Is there a way to re-install gcc 11.2?

here is the output to brew config and brew doctor

brew config
```
HOMEBREW_VERSION: 3.4.11
ORIGIN: https://github.com/Homebrew/brew
HEAD: 30e3d8883cde265cbc16ff94b49a3ac14137590e
Last commit: 12 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: b99460b09d3a7c7601df24d9fd5d7166d31e9616
Core tap last commit: 2 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.K3S9XrGGWt/org.xquartz:0
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 13.1.6 build 1316
Git: 2.32.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.4-arm64
CLT: 13.4.0.0.1.1651278267
Xcode: N/A
Rosetta 2: false
```

brew doctor
```
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  gcc@12

```

**Comment by SMillerDev** [comment]
Calling in @fxcoudert for GCC specific help

**Comment by fxcoudert** [comment]
So this looks like an issue at least one another person has been seeing: https://github.com/iains/gcc-11-branch/issues/3

In order to debug it, we would need some more information, ideally a minimal reproducer. Could you tell us:
- what you're compiling
- what is the command line that is producing this error
-  re-run the same and add the `-save-temps` flag, and send the intermediate files (`.i` and `.s`)

**Comment by sanjayg0** [comment]
I am going to try to construct a minimal example today since this is
occurring inside the build of a 300K+ line program -- though the compile
and link lines are modestly simple.

I will note that after several hours I located an error in the source
code.  The error, however, in my opinion should not have caused a link
error -- a compile warning or a runtime error, yes, but not a link error.

Here is the error in a nutshell:

0) subroutine sub1 that will be linked to a main program
1) Unintialized integer :: k
2) Local array real (kind=9) :: myarray(4)
3) Setting myarray(k) = myvalue -- with k never defined
4) Passing myarray to another subroutine call sub2(myarray)
5) Linking main sub1 and sub2 together throws the error.  main, sub1, and
sub2 all compile fine.

Minimally my compile lines looks like

gfortran -c -g -O2 -Wall sub1.f
gfortran -c -g -O2 -Wall sub2.f
gfortran -c -g -O2 -Wall fmain.f
gfortran fmain.o sub1.o sub2.o

Trying to make a minimal example, however, is proving to be hard because
-Wall is catching the uninitialized variable.  So it seems that the error
really stems from -Wuninitialized not seeing my uninitialized variable.
There are other features to the actual error that I have not described
above (global structures -- i.e. common blocks and some libraries).  I will
try to progressively make my minimal example more complex until it
reproduces the problem.

For what it is worth, I have attached the .s file from the compile of the
actual subroutine (called sld1d1u.f).  The symbol for the uninitialized
variable in the actual code is l (ell).  It compiled without warning even
though l (ell) is uninitialized, completely local, and -Wall was used.
 (No .i file was produced with the -save-temps flag).

-sanjay

**Comment by sanjayg0** [comment]
Here is my minimal example (attached zip file).  A bit of a mess but fully self-contained.  I left in all the IF checks but took out all the CALLs to other routines and all the computations except for one line (search for epsv, note in the real code the right-hand-side is properly initialized, only unil is uninitialized).

To build and see the error:

  gfortran -save-temps -c -g -O2 -Wall pjint2d.f
  gfortran -save-temps -c -g -O2 -Wall sld2d1u.f
  gfortran -save-temps -c -g -O2 -Wall fmain.f
  gfortran fmain.o sld2d1u.o pjint2d.o

The error occurs on link:

ld: in section __TEXT,__text reloc 13: symbol index out of range file 'sld2d1u.o' for architecture arm64
collect2: error: ld returned 1 exit status

The compile of sld2d1u.f does not catch the uninitialized variable "unil" and this appear to be the cause of the link error.  Note commenting out the call to pjint2d will remove the error.

Here is my -v on the compiler:
gfortran -v
Using built-in specs.
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/opt/homebrew/Cellar/gcc/11.3.0_1/bin/../libexec/gcc/aarch64-apple-darwin21/11/lto-wrapper
Target: aarch64-apple-darwin21
Configured with: ../configure --prefix=/opt/homebrew/opt/gcc --libdir=/opt/homebrew/opt/gcc/lib/gcc/11 --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-11 --with-gmp=/opt/homebrew/opt/gmp --with-mpfr=/opt/homebrew/opt/mpfr --with-mpc=/opt/homebrew/opt/libmpc --with-isl=/opt/homebrew/opt/isl --with-zstd=/opt/homebrew/opt/zstd --with-pkgversion='Homebrew GCC 11.3.0_1' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --build=aarch64-apple-darwin21 --with-system-zlib --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.3.0 (Homebrew GCC 11.3.0_1) 

[minex.zip](https://github.com/Homebrew/discussions/files/8750176/minex.zip)

Thanks in advance for looking at this,
-sanjay

**Reply by fxcoudert** [reply]
Thanks @sanjayg0 I can reproduce with this! Your help in reducing this to a minimal example will help fix the issue.

**Reply by sanjayg0** [reply]
Great.  If you need additional help, just ask.  We are very keen to see this fixed.

**Comment by iains** [comment]
* thanks, I can also repeat this
* interestingly, if I build with 12.1, the reduced case succeeds in linking (can you confirm/deny that from your end?)

**Reply by iains** [reply]
OK, so it looks like the GCC-11.3Dr0 output has a garbage relocation value for relocation 13, where (at least on my builds) the output is OK for GCC-11.2 and 12.1Dr0.  I am thinking that the workaround for https://github.com/iains/gcc-darwin-arm64/issues/3 might not be properly working around it.  However - it would be good to confirm the situation with 12.1 from someone else's build.

**Reply by fxcoudert** [reply]
I can confirm with the aarch64 Homebrew build from your gcc 12.1Dr0 that the reduced testcase does link without any error or warning. What I wonder is whether this is ARM-specific, or if it also occurs on Intel (I cannot test right now).

**Reply by iains** [reply]
yup
```
	adrp	x3, _epsv.3@PAGE-8	; 57	[c=4 l=4]  *movdi_aarch64/12
	str	xzr, [x3, #_epsv.3@PAGEOFF-8]	; 58	[c=4 l=4]  *movdf_aarch64/9
```
well, I guess that means https://github.com/iains/gcc-darwin-arm64/issues/3 is not fixed in Xcode, yet - and somehow the workaround has been messed up.

I'll try to figure this out over the next few days - note: we're up against a deadline for 10.4 and that's going to be an important release for macOS (last one GCC that can be built with a non-C++11 compiler).

**Reply by iains** [reply]
> I can confirm with the aarch64 Homebrew build from your gcc 12.1Dr0 that the reduced testcase does link without any error or warning. What I wonder is whether this is ARM-specific, or if it also occurs on Intel (I cannot test right now).

it does not occur on Intel, this is a bug in handling negative offsets on symbols (AFAIU it is acknowledged as a bug in Xcode for Arm64 and should be fixed at some point) - it just happens that clang does not generate this code pattern apparently....

**Reply by sanjayg0** [reply]
I will check later today with the reduced test case and my full code later today.  Will the .s files be helpful from the full case?

**Reply by iains** [reply]
I'd be most interested in whether your full code fails on 12.1 - the .s files are not needed (it is possible that there are two issues, but AFAICT the one described above was working on 12.1 with the reduced test case).

...  _The law of maximum cussedness does allow for two different bugs with the same symptom, of course_  ;)

**Comment by sanjayg0** [comment]
I can confirm that 12.1 will compile and link on the reduced test case (albeit missing the uninitialized).
However on the full test case, it fails with the error "in section __TEXT,__text reloc 573: symbol index our of range for architecture arm64".  I have attached the .s from the full code
[sld2d1u-fullcode-12.1.s.zip](https://github.com/Homebrew/discussions/files/8757981/sld2d1u-fullcode-12.1.s.zip)
.

**Reply by iains** [reply]
thanks, from that I can confirm that the problem is actually the same one - it's just that it manifests on the smaller case in 11.3 but only in the larger for 12.1.

@fxcoudert actually, there is no missed back-port .. the existing code to handle https://github.com/iains/gcc-darwin-arm64/issues/3 is in place (hence this affects 12.1 / 13.0 as well).  It seems that we have a situation in which improved optimisation is combining the instructions we carefully split during expand - so now I need to track down where that is occurring and place something to catch the case of combining a negative offset.

**Comment by sanjayg0** [comment]
In case it is helpful, here is the .s for the reduced test case from an  intel Mac using 11.3.  The compile again misses the uninitialized variable but will link.
[sld2d1u-reducedcase-11.3-intel.s.zip](https://github.com/Homebrew/discussions/files/8757994/sld2d1u-reducedcase-11.3-intel.s.zip)
.

**Reply by iains** [reply]
The fix for this is in GCC12.1-darwin-r1 and 11.3-darwin-r2 (releases imminent).

**Reply by sanjayg0** [reply]
Thanks!
